Add psalm linter for PHP

twitchard opened the pull request:
<!--
Before creating a pull request, do the following.

* Read the Contributing guide linked above first.
* Read the documentation that comes with ALE with `:help ale-development`.

Have fun!
-->

Submitted this previously in #1593, but I didn't follow through with the requested changes and it was closed for inactivity. This time, I added the documentation.

w0rp commented on ale_linters/php/psalm.vim:
```diff
@@ -0,0 +1,38 @@
+
+function! ale_linters#php#psalm#Handle(buffer, lines) abort
+    " Matches patterns like the following:
+
```
Remove the blank line here.

w0rp commented on ale_linters/php/psalm.vim:
```diff
@@ -0,0 +1,38 @@
+endfunction
+
+function! ale_linters#php#psalm#GetCommand(buffer) abort
+    return ale#Escape(ale_linters#php#psalm#GetExecutable(a:buffer)) . ' --diff --output-format=emacs %s'
```
You can write `command: '%e --diff --output-format=emacs %s`' below instead, and get the same result.

w0rp commented on ale_linters/php/psalm.vim:
```diff
@@ -0,0 +1,38 @@
+let g:ale_php_psalm_executable = get(g:, 'ale_php_psalm_executable', 'psalm')
+
+function! ale_linters#php#psalm#GetExecutable(buffer) abort
+    return ale#Var(a:buffer, 'php_psalm_executable')
```
You can write `'executable_callback': ale#VarFunc('php_psalm_executable')` below instead.

w0rp commented on ale_linters/php/psalm.vim:
```diff
@@ -0,0 +1,38 @@
+" Author: richard marmorstein <https://github.com/twitchard>
+" Description: plugin for Psalm, static analyzer for PHP
+
+let g:ale_php_psalm_executable = get(g:, 'ale_php_psalm_executable', 'psalm')
```
You can write `call ale#Set('php_psalm_executable', 'psalm')` instead.

w0rp commented on doc/ale-php.txt:
```diff
@@ -169,6 +169,16 @@ g:ale_php_phpstan_configuration               *g:ale_php_phpstan_configuration*
 
 
+===============================================================================
+psalm                                                           *ale-php-psalm*
```
Update the table of contents to include this heading too.

w0rp commented on ale_linters/php/psalm.vim:
```diff
@@ -0,0 +1,38 @@
+" Author: richard marmorstein <https://github.com/twitchard>
```
Add some tests for the command and executable here. See `:help ale-development`.

twitchard commented:
All right! I believe I've addressed all your comments.

I really admire your project by the way. The documentation and tooling around contributing is extremely impressive.

twitchard commented:
Anything else before this is ready to merge, @w0rp ?
